assembly/lunatic.ts

```diff
@@ -65,27 +82,342 @@ const receive_length_pointer = memory.data(sizeof<u32>());
+
+// @ts-ignore: valid external reference
+@external("lunatic", "sleep_ms")
+declare function sleep(ms: u64): void;
+
+
+@final export class Process {
+  private _pid: u32 = 0;
+  public get pid(): u32 { return this._pid; }
+
+  public static sleep(ms: u64): void {
+    sleep(ms);
+  }
+
+  /** This helper method spawns a Process with a simple boxed value of type T, must be integer or array. */
+  private static spawnWithBox<T>(val: T, callback: (val: T) => void): Process {
+
+    // box the callback and the value
+    let ptr = packCallbackWithValue(changetype<usize>(callback), val);
+
+    let threadCallback = (): void => {
+      // Get the payload from channel 0
+      let prepareResult = channel_receive_prepare(CHANNEL_INITIAL_PAYLOAD, receive_length_pointer);
+
+      // get the payload length and assert it's the correct size
+      let length = load<u32>(receive_length_pointer);
+      if (prepareResult == ChannelReceivePrepareResult.Fail) return;
+      assert(length == (sizeof<usize>() + sizeof<T>()));
+
+      // this is a static memory segment, allocated below __heap_base
+      let result = memory.data(sizeof<usize>() + sizeof<T>());
+
+      // obtain the static segment, callback, and val
+      channel_receive(result, length);
+      let callback = changetype<(val: T) => void>(load<usize>(result));
+      let val = load<T>(result, sizeof<usize>());
+
+      // start the thread
+      callback(val);
+    };
+
+    // send the box to the new thread
+    let t = new Process();
+    t._pid = spawn_with_context(
+      threadCallback.index,
+      ptr,
+      // packed message is the size of T + usize
+      sizeof<usize>() + sizeof<T>(),
+    );
+
+    // free the message pointer
+    heap.free(ptr);
+    return t;
+  }
+
+  /** This method spawns a process that receives an array. T is the value type, to aid in type detection. */
+  private static spawnWithArray<T>(val: Array<T>, callback: (val: Array<T>) => void): Process {
+    // private buffer: ArrayBuffer;
+    // private dataStart: usize;
+    // private byteLength: i32;
+    // private _length: i32;
+
+    // we can obtain the dataStart and byteLength value doing a manual load here
+    let dataStart = load<usize>(changetype<usize>(val), offsetof<Array<T>>("dataStart"));
+    let byteLength = load<i32>(changetype<usize>(val), offsetof<Array<T>>("byteLength"));
+
+    let threadCallback = (): void => {
+      // Get the payload and the length
+      let prepareResult = channel_receive_prepare(CHANNEL_INITIAL_PAYLOAD, receive_length_pointer);
+      if (prepareResult == ChannelReceivePrepareResult.Fail) return;
+      let length = load<u32>(receive_length_pointer);
+
+      let messagePointer = heap.alloc(length);
+
+      channel_receive(messagePointer, length);
+
+      let callback = changetype<(val: Array<T>) => void>(load<usize>(messagePointer));
+      let byteLength = length - sizeof<usize>();
+      let arrayLength = byteLength >>> alignof<T>();
```

Comment:
I wonder if this always works, since `.length` on an array can be set in a way that it doesn't necessarily correspond to the backing buffer's `.byteLength` (capacity). Please excuse if this is handled and I overlooked something, just briefly glimpsed over it :)

Reply:
Yeah that sounds right! I'll have to check how the `length` property works.  Thanks for the review. I might have missed this.

Reply:
```ts
    let byteLength = load<i32>(changetype<usize>(val), offsetof<Array<T>>("length_")) << alignof<T>();
```

This works. I forgot that there was a computed `length_` property.